File: src/xianyu_cli/utils/credential.py
"""Credential storage with TTL management.

Credentials are saved to ~/.config/xianyu-cli/credential.json with a timestamp.
Stale credentials (beyond TTL) trigger re-authentication.
"""

from __future__ import annotations

import json
import os
import stat
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from xianyu_cli.utils._common import (
    CREDENTIAL_FILE,
    DEFAULT_CREDENTIAL_TTL_HOURS,
    ensure_config_dir,
)


class Credential:
    """Represents a saved authentication credential."""

    def __init__(
        self,
        cookies: dict[str, str],
        user_id: str = "",
        nickname: str = "",
        source: str = "unknown",
        saved_at: str | None = None,
    ):
        self.cookies = cookies
        self.user_id = user_id
        self.nickname = nickname
        self.source = source
        self.saved_at = saved_at or datetime.now(timezone.utc).isoformat()

    def is_expired(self, ttl_hours: int = DEFAULT_CREDENTIAL_TTL_HOURS) -> bool:
        """Check if the credential has exceeded its TTL."""
        try:
            saved = datetime.fromisoformat(self.saved_at)
            if saved.tzinfo is None:
                saved = saved.replace(tzinfo=timezone.utc)
            age_hours = (datetime.now(timezone.utc) - saved).total_seconds() / 3600
            return age_hours > ttl_hours
        except (ValueError, TypeError):
            return True

    def has_session(self) -> bool:
        """Check if the credential has core session cookies (unb = logged in)."""
        return bool(self.cookies.get("unb"))

    @property
    def m_h5_tk(self) -> str | None:
        """Extract token from _m_h5_tk cookie (part before underscore)."""
        tk = self.cookies.get("_m_h5_tk", "")
        if "_" in tk:
            return tk.split("_")[0]
        return tk or None

    def to_dict(self) -> dict[str, Any]:
        return {
            "version": 1,
            "saved_at": self.saved_at,
            "source": self.source,
            "user_id": self.user_id,
            "nickname": self.nickname,
            "cookies": self.cookies,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Credential:
        return cls(
            cookies=data.get("cookies", {}),
            user_id=data.get("user_id", ""),
            nickname=data.get("nickname", ""),
            source=data.get("source", "unknown"),
            saved_at=data.get("saved_at"),
        )


def save_credential(cred: Credential, path: Path | None = None) -> None:
    """Save credential to disk with restricted file permissions (0600)."""
    path = path or CREDENTIAL_FILE
    ensure_config_dir()
    data = json.dumps(cred.to_dict(), ensure_ascii=False, indent=2)
    path.write_text(data, encoding="utf-8")
    os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)


def load_credential(path: Path | None = None) -> Credential | None:
    """Load credential from disk, returning None if not found."""
    path = path or CREDENTIAL_FILE
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return Credential.from_dict(data)
    except (json.JSONDecodeError, KeyError):
        return None


def delete_credential(path: Path | None = None) -> bool:
    """Delete saved credential file. Returns True if file was deleted."""
    path = path or CREDENTIAL_FILE
    if path.exists():
        path.unlink()
        return True
    return False

File: golang/utils/credential.go
package utils

import (
	"encoding/json"
	"os"
	"strings"
	"time"
)

// Credential represents saved authentication credentials.
type Credential struct {
	Version  int               `json:"version"`
	SavedAt  string            `json:"saved_at"`
	Source   string            `json:"source"`
	UserID   string            `json:"user_id"`
	Nickname string            `json:"nickname"`
	Cookies  map[string]string `json:"cookies"`
}

// NewCredential creates a new credential with current timestamp.
func NewCredential(cookies map[string]string, source string) *Credential {
	return &Credential{
		Version: 1,
		SavedAt: time.Now().UTC().Format(time.RFC3339),
		Source:  source,
		Cookies: cookies,
	}
}

// IsExpired checks if the credential has exceeded its TTL.
func (c *Credential) IsExpired(ttlHours ...int) bool {
	ttl := DefaultCredentialTTLHours
	if len(ttlHours) > 0 {
		ttl = ttlHours[0]
	}
	saved, err := time.Parse(time.RFC3339, c.SavedAt)
	if err != nil {
		// Try ISO format without timezone
		saved, err = time.Parse("2006-01-02T15:04:05", c.SavedAt)
		if err != nil {
			return true
		}
		saved = saved.UTC()
	}
	age := time.Since(saved)
	return age.Hours() > float64(ttl)
}

// HasSession checks if the credential has core session cookies.
func (c *Credential) HasSession() bool {
	return c.Cookies["unb"] != ""
}

// MH5TK extracts token from _m_h5_tk cookie (part before underscore).
func (c *Credential) MH5TK() string {
	tk := c.Cookies["_m_h5_tk"]
	if idx := strings.Index(tk, "_"); idx > 0 {
		return tk[:idx]
	}
	return tk
}

// SaveCredential saves credential to disk with restricted permissions (0600).
func SaveCredential(cred *Credential, path ...string) error {
	p := CredentialFile
	if len(path) > 0 {
		p = path[0]
	}
	if err := EnsureConfigDir(); err != nil {
		return err
	}
	data, err := json.MarshalIndent(cred, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(p, data, 0600)
}

// LoadCredential loads credential from disk.
func LoadCredential(path ...string) *Credential {
	p := CredentialFile
	if len(path) > 0 {
		p = path[0]
	}
	data, err := os.ReadFile(p)
	if err != nil {
		return nil
	}
	var cred Credential
	if err := json.Unmarshal(data, &cred); err != nil {
		return nil
	}
	return &cred
}

// DeleteCredential deletes saved credential file.
func DeleteCredential(path ...string) bool {
	p := CredentialFile
	if len(path) > 0 {
		p = path[0]
	}
	if err := os.Remove(p); err != nil {
		return false
	}
	return true
}
